### PR TITLE
[PWX-34708] Support multi-DR for normal application resources and CRs with suspend options

### PR DESCRIPTION
**Pull request by olavangad-px**
**What type of PR is this?**
>bug

**What this PR does / why we need it**:
[PWX-34708](https://portworx.atlassian.net/browse/PWX-34708)
To carry forward the appropriate replicas value in case of multi-DR
For now we have only fixed the bug for normal application resources and CRs with suspend options defined.
Changes for CRs with stashStrategy enabled will be taken up separately.

**Does this PR change a user-facing CRD or CLI?**:
No

**Is a release note needed?**:
Yes, TBD

**Does this change need to be cherry-picked to a release branch?**:
Yes, 23.11.0

**Testing Details**
Have manually tested my changes in a setup of 3 clusters with ASYNC-ASYNC relationship. (Metro or Async should not matter from code point of view)
Have tested that correct annotation is carried forward for following cases -> 

1. Deployment with replicas set to 2, is migrated from C1 to C2. And then from C2 to C3.
2. MongoDBCommunity resource with spec.members set to 3, is migrated from C1 to C2. And then from C2 to C3. (suspend path spec.members with suspend.Type int)
3. Kubevirt virtualmachine with runStrategy set to Always , is migrated from C1 to C2. And then from C2 to C3. (suspend path spec.runStrategy with suspend.Type string and suspend.Value Halted)

Have verified that created resources have expected annotation values in both C2 and C3.

[PWX-34708]: https://portworx.atlassian.net/browse/PWX-34708?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: pkg/migration/controllers/migration.go [hunk 1]
```diff
@@ -1547,18 +1567,55 @@ func (m *MigrationController) prepareCRDClusterResource(
 					return fmt.Errorf("unable to find suspend path, err: %v", err)
 				}
 				disableVersion = int64(0)
+				if curr == 0 && suspendAnnotationIsPresent {
+					//suspendAnnotation has value set as {currVal + "," + suspend.Value}
+					//we need to extract only the currVal from it
+					annotationValue := strings.Split(SuspendAnnotationValue, ",")[0]
```
Comment (diptiranjanpx) [inline pkg/migration/controllers/migration.go:1573]: Is it really required to do the split and again modify the `curr` value and later again add it to the annotation?

If curr == 0 &&  suspendAnnotationIsPresent 
       We don't have to set anything, we can take the object with the annotation itself.

File: pkg/migration/controllers/migration.go [hunk 2]
```diff
@@ -1547,18 +1567,55 @@ func (m *MigrationController) prepareCRDClusterResource(
...
+						}
+					}
+				}
 				currVal = curr
 			} else {
 				return fmt.Errorf("invalid type %v to suspend cr", suspend.Type)
 			}
 
+			//No need to scale down if StartApplications is set to True in the migration spec
+			if *migration.Spec.StartApplications {
```
Comment (diptiranjanpx) [inline pkg/migration/controllers/migration.go:1614]: Why is this required, can you please explain?

**Review by pp511 — COMMENTED** [review]
Did take one pass at it. Looks fine so far.
Will take a second look at it again tomorrow and  will dry-run with a concrete example to make sure everything is ok.

File: pkg/migration/controllers/migration.go (hunk 1 above)
Comment (olavangad-px) [reply pkg/migration/controllers/migration.go:1573]: 1. The condition I am checking as of now is curr == 0 && suspendAnnotation is > 0. So I need to extract the annotation value for that.
2. Additionally I am also updating the suspend path value with the annotationValue in line #1583
This is done for the scenario when migration.spec.startApplications is true.

Let's consider the case where resources were migrated from C1 to C2. Now the resource in C2 is scaled down so actual replica value will be 0 while annotation will have the replicaCount from C1 (let's say it is 3)
Now if I migrate the resource from C2 to C3 with startApplications as true, I would not be scaling down the resource and migrating it as is. For this case I should have modified the actual replica value in the resource as well. This is what is done in #1583

So that in #1614 when we return nil because startApplications is true and we do not need to scale down, the resource applied in C3 will have replicaCount set to 3.

File: pkg/migration/controllers/migration.go (hunk 2 above)
Comment (olavangad-px) [reply pkg/migration/controllers/migration.go:1614]: In the existing implementation this check is added at the beginning of the functions (#1519 in previous code) since we do not need to scale down the resource if we are going start the applications in the target cluster.

Since as part of this bug, we may want a different value of replica compared to the existing count (which could be 0 and the real useful value is stored in the annotation), I have moved the check to a different place.

I have explained how we get the appropriate value for this in the above comment with an example.

File: pkg/migration/controllers/migration.go (hunk 1 above)
Comment (diptiranjanpx) [reply pkg/migration/controllers/migration.go:1573]: Got it thanks, this is required for `startApplications : true`. I did not think about that.

**Review by diptiranjanpx — APPROVED** [review]

**Review by pp511 — APPROVED** [review]
Lgtm
